runtime/src/comm/ugni/comm-ugni.c

```diff
@@ -2507,164 +2541,198 @@ void chpl_comm_rollcall(void)
 }
 
 
-static void make_shared_heap(void)
+static
+void ensure_registered_heap_info_set(void)
 {
-  assert(!registered_heap_info_set);
+  if (!registered_heap_info_set
+      && pthread_once(&registered_heap_once_flag, make_registered_heap) != 0) {
+    CHPL_INTERNAL_ERROR("pthread_once(&registered_heap_once_flag) failed");
+  }
+}
 
-  if (chpl_numNodes == 1) {
-    using_hugepages = false;
-    registered_heap_start = NULL;
+
+static
+void make_registered_heap(void)
+{
+  size_t size_from_env;
+
+  ensure_hugepage_info_set();
+
+  if (!using_hugepages
+      || getenv("HUGETLB_MORECORE") == NULL
+      || (size_from_env = chpl_comm_getenvMaxHeapSize()) == 0) {
     registered_heap_size  = 0;
+    registered_heap_start = NULL;
     registered_heap_info_set = 1;
+    atomic_thread_fence(memory_order_release);
     return;
   }
 
-  using_hugepages = (getenv("HUGETLB_MORECORE") == NULL) ? false : true;
-  if (using_hugepages) {
-    //
-    // If the heap is supposed to be on hugepages, acquire that space
-    // now.
-    //
-    const size_t page_size = hugepage_size = gethugepagesize();
-    const size_t nic_max_pages = (size_t) 1 << 14; // not publicly defined
-    const size_t nic_max_mem = nic_max_pages * page_size;
-    size_t nic_allowed_mem;
-    size_t max_heap_size;
-    size_t size;
-    size_t decrement;
-    void*  start;
+  //
+  // The heap is supposed to be of fixed size and on hugepages.  Set
+  // it up.  (If it's to be dynamically extensible, that's handled
+  // separately through chpl_comm_regMem*().)
+  //
+  const size_t nic_max_pages = (size_t) 1 << 14; // not publicly defined
+  const size_t nic_max_mem = nic_max_pages * hugepage_size;
```

Comment:
I have a really hard time following when hugepage_size is set or not. It might be easier to follow if it was a static inline function instead of a global, but if you're confident that it's set when it needs to be, that's good enough for me

Reply:
Done, in e3507715.